mut/stage.py

```diff
@@ -52,21 +51,219 @@
+        flag = 'C' if new_file else 'M'
+        self.commands_upload.append((flag, path, key))
+
+    def redirect(self, from_key: str, to_url: str) -> None:
+        """Create an S3 redirect."""
+        self.commands_redirect.append((from_key, to_url))
+
+    def print(self) -> None:
+        """Print to stdout all actions that will be taken by ChangeSet.commit()."""
+        files_deleted = 0
+        redirects_deleted = 0
+        files_modified = 0
+        files_created = 0
+
+        for command in self.commands_upload:
+            flag, _, key = command
+            if flag is 'C':
+                files_created += 1
+            elif flag is 'M':
+                files_modified += 1
+            else:
+                raise ValueError('Unknown upload flag {}'.format(repr(flag)))
+
+            print('{}  {}'.format(flag, key))
+
+        for redirect in self.commands_redirect:
+            print('R  {} -> {}'.format(redirect[0], redirect[1]))
+
+        for deletion in self.commands_delete:
+            flag, key = deletion
+            if flag is 'D':
+                files_deleted += 1
+            elif flag is 'DR':
+                redirects_deleted += 1
+            else:
+                raise ValueError('Unknown deletion flag {}'.format(repr(flag)))
+
+            print('{:<2} {}'.format(flag, key))
+
+        print('\nSummary\n=======')
+        print('Files Deleted:     {}'.format(files_deleted))
+        print('Redirects Deleted: {}'.format(redirects_deleted))
+        print('Files Modified:    {}'.format(files_modified))
+        print('Files Created:     {}'.format(files_created))
+        print('Redirects Created: {}'.format(len(self.commands_redirect)))
+
+    def commit(self, s3: Any) -> None:
+        """Apply the set of operations stored in this instance."""
+        tasks = []
+        for command in self.commands_upload:
+            _, src_path, key = command
+            task = functools.partial(self.__upload, s3, src_path, key)
+            tasks.append(cast(Callable[[None], None], task))
+
+        for redirect in self.commands_redirect:
+            src, dest = redirect
+            task = functools.partial(self.__redirect, s3, src, dest)
+            tasks.append(cast(Callable[[None], None], task))
+
+        run_pool(tasks)
+
+        # S3 caps delete requests to 1,000 keys.
+        for chunk in chunks(self.commands_delete, 999):
+            s3.delete_objects(Delete={
+                'Objects': [{'Key': key} for _, key in chunk],
+                'Quiet': True
+            })
+
+    def __upload(self, s3: Any, src_path: str, key: str) -> None:
+        """Thread worker helper to handle uploading a single file to S3."""
+        try:
+            s3.upload_file(src_path, key, ExtraArgs={
+                'StorageClass': 'REDUCED_REDUNDANCY',
+                'ContentType': mimetypes.guess_type(src_path)[0] or 'binary/octet-stream'
+            })
+            sys.stdout.write('.')
```

Comment:
Is this where the ............... is printed to the console?  Is this one '.' per file? If so, I feel like this can get very unwieldy for bigger docs properties like the manual.

Reply:
It is. Remember though that in the print() stage, each file will be printed on its own line. This doesn't even compare in its unwieldiness.